sdk/src/main/java/com/google/cloud/dataflow/sdk/coders/Proto2Coder.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.dataflow.sdk.coders;
+
+import com.google.protobuf.ExtensionRegistry;
+import com.google.protobuf.GeneratedMessage.GeneratedExtension;
+import com.google.protobuf.Message;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+
+/**
+ * An encoder using Google Protocol Buffers 2 binary format.
+ * <p>
+ * To learn more about Protocol Buffers, visit:
+ * https://developers.google.com/protocol-buffers
+ * <p>
+ * To use, specify the {@code Coder} type on a PCollection:
+ * <pre>
+ * {@code
+ * PCollection<MyProtoMessage> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProtoMessage.class));
+ * }
+ * </pre>
+ * <p>
+ * Custom message extensions are also supported, but the coder must be made
+ * aware of them explicitly:
+ * <pre>
+ * {@code
+ * PCollection<MyProtoMessage> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProtoMessage.class)
+ *              .withExtensions(MyProtoMessage.extension1, MyProtoMessage.extension2));
+ * }
+ * </pre>
+ *
+ * @param <T> the type of elements handled by this coder, must extend {@code Message}
+ */
+public class Proto2Coder<T extends Message> extends StandardCoder<T> {
+
+    public static <T extends Message> Proto2Coder<T> of(Class<T> protoMessageClass) {
+      return new Proto2Coder<>(protoMessageClass);
+    }
+
+    private final Class<T> protoMessageClass;
+    private final ExtensionRegistry extensionRegistry = ExtensionRegistry.newInstance();
```

Comment:
StandardCoder needs to support serialization to/from CloudObjects via the overridable asCloudObject() and the @JsonCreator annotated method of(...)

See for an example: 
https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/coders/AvroCoder.java#L118
https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/coders/AvroCoder.java#L165

Also, the protoMessageClass and extensionRegistry are not portable beyond Java.
To make this coder portable you will need to provide a cloud object mapping which is able to reconstruct the FileDescriptor. I would suggest the following format:
```json
{
  "message_type": "*the descriptors name*",
  "file_descriptor_proto": "*the file descriptor proto serialized*",
  "dependencies": [
     {
       "file_descriptor_proto": "*dependency file descriptor proto*",
       "dependencies": [ recursively more dependencies ],
     }
     ... more dependencies ...
  ]
}
```

You will then be able to reconstruct the FileDescriptor proto using:
https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/Descriptors.FileDescriptor#buildFrom(com.google.protobuf.DescriptorProtos.FileDescriptorProto, com.google.protobuf.Descriptors.FileDescriptor[])
And then get back the Descriptor using:
https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/Descriptors.FileDescriptor#findMessageTypeByName(java.lang.String)

From then on you should be able to use the Descriptor to create messages.